[libc] Make use of 32-bit time_t a config option

statham-arm opened the pull request:
The 32-bit Arm builds of libc define time_t to be `__INTPTR_TYPE__`, i.e. a 32-bit integer. This is commented in the commit introducing it (75398f28ebdb600) as being for compatibility with glibc. But in the near future not even every AArch32 build of glibc will have a 32-bit time_t: Debian is planning that their next release (trixie) will have switched to 64-bit. And non-Linux builds of this libc (e.g. baremetal) have no reason to need glibc compatibility in the first place – and every reason _not_ to want to start using a 32-bit time_t in 2024 or later.

So I've replaced the `#ifdef` in `llvm-libc-types/time_t.h` with two versions of the header file, chosen in `CMakeLists.txt` via a new configuration option. This involved adding an extra parameter to the cmake `add_header` function to specify different names for the header file in the source and destination directories.

statham-arm committed:
[libc] Make use of 32-bit time_t a config option

The 32-bit Arm builds of libc define time_t to be `__INTPTR_TYPE__`,
i.e. a 32-bit integer. This is commented in the commit introducing it
(75398f28ebdb600) as being for compatibility with glibc. But in the
near future not even every AArch32 build of glibc will have a 32-bit
time_t: Debian is planning that their next release (trixie) will have
switched to 64-bit. And non-Linux builds of this libc (e.g. baremetal)
have no reason to need glibc compatibility in the first place – and
every reason _not_ to want to start using a 32-bit time_t in 2024 or
later.

So I've replaced the `#ifdef` in `llvm-libc-types/time_t.h` with two
versions of the header file, chosen in `CMakeLists.txt` via a new
configuration option. This involved adding an extra parameter to the
cmake `add_header` function to specify different names for the header
file in the source and destination directories.

lntue commented on libc/cmake/modules/LLVMLibCHeaderRules.cmake:
```diff
@@ -21,7 +21,11 @@ function(add_header target_name)
   file(RELATIVE_PATH relative_path ${LIBC_INCLUDE_SOURCE_DIR} ${absolute_path})
   set(dest_file ${LIBC_INCLUDE_DIR}/${relative_path})
-  set(src_file ${CMAKE_CURRENT_SOURCE_DIR}/${ADD_HEADER_HDR})
+  if(ADD_HEADER_SRCHDR)
```
I think the whole project is kind of used to `HDR/HDRS/SRCS` are the source files.  Can you change the argument to something like `DEST_HDR`, and update the `dest_file`/`absolute_path` instead?

lntue commented on libc/include/llvm-libc-types/CMakeLists.txt:
```diff
@@ -58,7 +58,16 @@ add_header(pthread_rwlock_t HDR pthread_rwlock_t.h DEPENDS .__futex_word .pid_t)
 add_header(pthread_t HDR pthread_t.h DEPENDS .__thread_type)
 add_header(rlim_t HDR rlim_t.h)
-add_header(time_t HDR time_t.h)
+if(LIBC_CONF_TIME_64BIT)
```
nit:
```
if(LIBC_TARGET_ARCHITECTURE_IS_ARM AND NOT LIBC_CONF_TIME_64BIT)
  # Set time_t to 32-bit for compatibility with glibc
  add_header(time_t HDR time_t_32.h DEST_HDR time_t.h)
else
  # Use 64-bit time_t
  add_header(time_t HDR time_t_32.h DEST_HDR time_t.h)
endif()
```

lntue reviewed (commented):
This way of selecting headers to be copying over for public includes is probably cleanest for full build, but it will be a bit troublesome for proxy-headers and other headers inside `libc/include` that directly include `llvm-libc-types/time_t.h`.

For the proxy-header `libc/src/types/time_t.h`, you can use some flag like `LIBC_TYPES_TIME_T_IS_64_BIT` and `LIBC_TYPES_TIME_T_IS_32_BIT` to include the correct ones, and define it in common flags for objects:  inside the if at https://github.com/llvm/llvm-project/blob/main/libc/cmake/modules/LLVMLibCCompileOptionRules.cmake#L86 and similar one in the same file for tests.

For the other problem that other headers inside `libc/include/` also directly include `time_t`, I can think of a solution, that is directly copy either `time_t_64.h`/`time_t_32.h` to `libc/include/llvm-libc-types/time_t.h` instead.

And I think this might be a simpler solution: is you do the selection and copy to `time_t.h` in the same folder, before `add_header(time_t HDR time_t.h DEPENDS <copy_command>)`.  And in that case, you will not need to update proxy headers or add extra compile options.  WDYT?

lntue commented:
Also, you might one to just leave a copy of `libc/include/llvm-libc-types/time_t.h` (same content as `time_t_64.h`) so that it won't break downstream projects with different build systems.

statham-arm commented:
> For the other problem that other headers inside `libc/include/` also directly include `time_t`, I can think of a solution, that is directly copy either `time_t_64.h`/`time_t_32.h` to `libc/include/llvm-libc-types/time_t.h` instead.
> 
> And I think this might be a simpler solution: is you do the selection and copy to `time_t.h` in the same folder, before `add_header(time_t HDR time_t.h DEPENDS <copy_command>)`. And in that case, you will not need to update proxy headers or add extra compile options. WDYT?

To check I've understood correctly, do you mean copying one of my two time_t header files to the path `libc/include/llvm-libc-types/time_t.h` in the _source_ tree, rather than the build directory?

I'd rather avoid that if possible, because it's often useful to be able to do multiple out-of-tree builds from the same source directory with different configurations. So involving the source tree in your build in any way that is not read-only is dangerous.

(I was already a bit culture-shocked by the part of the libc build that auto-updated `libc/docs/configure.rst` when I edited `config.json`! But that only violates the _letter_ of this principle, not the spirit: the change it made in the source directory isn't specific to any build configuration.)

But I take your point about headers in the source `libc/include` subdir transitively including each other in a way that doesn't pay attention to my selection operation in the build dir. Certainly there needs to be _some_ answer to that.

When I first started looking at this problem, my first guess was that there would be an autogenerated header file called something like `llvm-libc-config.h`, similar to `libcxx/include/__config_site`, where you could write macro definitions of things discovered at cmake configure time. Then I could have written something into that file like `#define __LIBC_COPT_TIME_T_IS_64_BIT`, and `time_t.h` could have checked that ifdef instead of `__arm__` (or perhaps as well) to decide on its time_t definition, and everyone would be happy -- that macro definition would be available to the libc build _and_ to user applications including the headers. But as far as I can see there is no such config header in libc.

That said: I don't think I personally have any use case for wanting to do multiple out-of-tree builds from the same source directory with different sizes of time_t. I do want to do lots of builds from the same source dir, but they're all embedded, and I'll want all of them to have 64-bit time_t. So if you really think writing to the source dir is the best thing in this case, I don't think it'll inconvenience me _too_ much.

lntue commented:
I agree about the danger of modifying files in the source tree with build actions.  In that case, I think the following would solve our problems:
- Copy either `time_t_32.h` or `time_t_64.h` to public include `time_t.h` as in your current change.
- Keep the current `libc/include/llvm-libc-types/time_t.h`, adding the condition `&& !defined(LIBC_TYPES_TIME_T_64_BIT)` in addition to current arm target detection.
- Add `-DLIBC_TYPES_TIME_T_64_BIT` to common object and test options in full build mode if `LIBC_CONF_TIME_64BIT` is on.
WDYT?

statham-arm committed:
Review feedback: rework how the right time_t header is selected.

Now the decision results in LIBC_TYPES_TIME_T_IS_32_BIT being set in
the cmake scripts. That controls which of time_t_{32,64}.h is copied
to the output header time_t.h. But it also sets a flag on the compile
command lines, which cause hdr/llvm-libc-types/time_t.h to #include
the same one of those headers at library build time.

statham-arm commented:
That sounds reasonable to me. I've pushed a patch that implements this, I think.

I've flipped it round so that the `#define` is used for a 32-bit time_t. That way the common case (and hopefully more common as we approach Y2038) is the default one, without needing an extra command-line option.

And I've centralised the cmake code that decides on the size of time_t, so that the same variable is used to control both the command-line `#define` and the choice of output `time_t.h`.

lntue commented on libc/cmake/modules/LLVMLibCCompileOptionRules.cmake:
```diff
@@ -71,6 +71,10 @@ function(_get_compile_options_from_config output_var)
     list(APPEND config_options "-DLIBC_QSORT_IMPL=${LIBC_CONF_QSORT_IMPL}")
   endif()
 
+  if(LIBC_TYPES_TIME_T_IS_32_BIT)
```
We should only pass the compile flag in full build mode:
```
if(LIBC_TYPES_TIME_T_IS_32_BIT AND LLVM_LIBC_FULL_BUILD)
```

lntue commented on libc/include/llvm-libc-types/time_t_32.h:
```diff
@@ -0,0 +1,14 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_TYPES_TIME_T_H
```
This file and `time_t_64.h` are both using the same header guard as `time_t.h`, `time_t` will not be defined when included from `time_t.h`.

statham-arm replied:
Good point – which must mean that nothing at all in the build is _currently_ using this `time_t.h` to redirect to the right sub-header, or else my own test builds would have failed!

lntue replied:
Did you try with full build mode?  I found out the header guard issue from failed tests when trying your PR in full build mode.

statham-arm replied:
Ah, it was in the tests? That explains it.

I'm _mostly_ using full build mode, because I'm testing this libc on bare metal. But I haven't figured out how to run the tests in that setup yet, so that's why I didn't spot this. Good thing you did.

lntue replied:
Yes, it's good that we can exercise the same code paths on multiple targets.  It does help to find hidden issues.

statham-arm committed:
Review nits: disambiguate guard macros, conditionalise -D

nickdesaulniers commented:
I suspect that we could just force `time_t` to be 64b always on 32b ARM as part of our own ABI.
